### PR TITLE
initial cut at a room summary API

### PR DESCRIPTION
**Pull request by ara4n**
this works, although considers all room membership events rather than just joined
ones as i'm failing to see a nice way to filter based on membership.

Implements matrix-org/matrix-doc#688

~~Builds on #3568 (in turn #3567 (in turn #3331 (in turn #2970))).~~ (all now merged)

Sytest at https://github.com/matrix-org/sytest/pull/470

### CODE REVIEW

File: synapse/handlers/sync.py [hunk 1]
```diff
@@ -494,10 +495,105 @@ def get_state_at(self, room_id, stream_position, types=None, filtered_types=None
...
+                if typ == EventTypes.Member
+            }
+
+            for ev in batch.events:
+                if ev.type == EventTypes.Member:
+                    existing_members[ev.state_key] = True
+
+            missing_hero_event_ids = [
+                member_ids[hero_id]
+                for hero_id in summary['heros']
```
Comment (bwindels) [inline synapse/handlers/sync.py]: Should be `m.heros`? I get a `exceptions.KeyError: 'heros'` when running this locally.

Comment (bwindels) [reply synapse/handlers/sync.py]: Fixed it, as am 99% certain that that is what it should be, and it unblocked me locally from syncing.

**Comment by bwindels** [comment]
`m.joined_member_count` seems to be irrespective of membership. While testing in a DM, upon the other user leaving, I get this from /sync for the given room. Is it also intended that the left user is still included in `m.heros`? I guess it could, as you won't want the room name to change, but different from what the proposal says:

```
{
  "!VHGOLZdMXhOiFmizWb:localhost": {
    "timeline": {
      "limited": false,
      "prev_batch": "s500_5160_4_83_232_1_1_56_1",
      "events": [
        {
          "origin_server_ts": 1532351680209,
          "sender": "@bruno5:localhost",
          "event_id": "$153235168010hSmVV:localhost",
          "unsigned": {
            "prev_content": {
              "membership": "join",
              "avatar_url": "mxc://localhost/JyLQUEVWImjlOIyJRLGxqZtD",
              "displayname": "Bruno Mars"
            },
            "prev_sender": "@bruno5:localhost",
            "replaces_state": "$15323515718rGSwF:localhost",
            "age": 81
          },
          "state_key": "@bruno5:localhost",
          "content": {
            "membership": "leave"
          },
          "membership": "leave",
          "type": "m.room.member"
        }
      ]
    },
    "summary": {
      "m.joined_member_count": 2,
      "m.heros": [
        "@bruno1:localhost",
        "@bruno5:localhost"
      ]
    }
  }
}
```

File: synapse/handlers/sync.py [hunk 2]
```diff
@@ -494,10 +495,125 @@ def get_state_at(self, room_id, stream_position, types=None, filtered_types=None
...
+
+        for ev in member_events.values():
+            if ev.content.get("membership") == Membership.JOIN:
+                joined_user_ids.append(ev.state_key)
+            elif ev.content.get("membership") == Membership.INVITE:
+                invited_user_ids.append(ev.state_key)
+
+        summary["m.joined_member_count"] = len(joined_user_ids)
+        if invited_user_ids:
+            summary["m.invited_member_count"] = len(invited_user_ids)
```
Comment (bwindels) [inline synapse/handlers/sync.py]: Wouldn't this only not send it when the count is 0? Whereas the MSC mentions that not sending it means it hasn't changed since last time it was sent?

File: synapse/handlers/sync.py [hunk 3]
```diff
@@ -1470,6 +1586,15 @@ def _generate_room_entry(self, sync_result_builder, ignored_users,
             full_state=full_state
         )
 
+        summary = None
```
Comment (manuroe) [inline synapse/handlers/sync.py]: Should be `summary = {}`.

File: synapse/handlers/sync.py [hunk 4]
```diff
@@ -1470,6 +1586,15 @@ def _generate_room_entry(self, sync_result_builder, ignored_users,
...
+        if (
+            sync_config.filter_collection.lazy_load_members() and
+            any(ev.type == EventTypes.Member for ev in batch.events)
```
Comment (manuroe) [inline synapse/handlers/sync.py]: This seems to not handle the case of an initial sync where `batch.events` may contain no membership events.

Comment (bwindels) [inline synapse/handlers/sync.py]: Also calculate the summary on initial sync?

File: synapse/storage/_base.py [hunk 5]
```diff
@@ -1153,8 +1153,7 @@ def get_user_count_txn(self, txn):
         """
         sql_count = "SELECT COUNT(*) FROM users WHERE is_guest = 0;"
         txn.execute(sql_count)
-        count = txn.fetchone()[0]
-        defer.returnValue(count)
+        return txn.fetchone()[0]
```
Comment (richvdh) [inline synapse/storage/_base.py]: grr. can you fix the docstring while you are there?

File: synapse/handlers/sync.py (hunk 2 above)
Comment (ara4n) [reply synapse/handlers/sync.py]: oops, good point. have changed it to send them every time for now (including if 0); the optimisation to only send them when they change can be done later.

File: synapse/handlers/sync.py [hunk 6]
```diff
@@ -494,10 +495,125 @@ def get_state_at(self, room_id, stream_position, types=None, filtered_types=None
...
+            if (
+                summary["m.joined_member_count"] == 0 and
+                summary["m.invited_member_count"] == 0
+            ):
+                summary['m.heroes'] = sorted(
+                    [user_id for user_id in member_ids.keys() if user_id != me]
+                )[0:5]
+            else:
+                summary['m.heroes'] = sorted(
+                    [user_id for user_id in joined_user_ids if user_id != me]
```
[the remark this answers is not in the document]
Comment (ara4n) [reply synapse/handlers/sync.py]: oops, fixed.

File: synapse/handlers/sync.py (hunk 4 above)
Comment (ara4n) [reply synapse/handlers/sync.py]: oops, fixed.

File: synapse/handlers/sync.py [hunk 7]
```diff
@@ -494,10 +495,125 @@ def get_state_at(self, room_id, stream_position, types=None, filtered_types=None
...
+            for (t, state_key), event_id in state_ids.iteritems()
+            if t == EventTypes.Member
+        }
+        name_id = state_ids.get((EventTypes.Name, ''))
+        canonical_alias_id = state_ids.get((EventTypes.CanonicalAlias, ''))
+
+        summary = {}
+
+        # FIXME: it feels very heavy to load up every single membership event
+        # just to calculate the counts. get_joined_users_from_context might
```
[the remark this answers is not in the document]
Comment (ara4n) [reply synapse/handlers/sync.py]: um, ok

File: synapse/handlers/sync.py [hunk 8]
```diff
@@ -494,10 +495,125 @@ def get_state_at(self, room_id, stream_position, types=None, filtered_types=None
...
+            cache = self.lazy_loaded_members_cache.get(cache_key)
+
+            existing_members = {
+                user_id: True for (typ, user_id) in state.keys()
+                if typ == EventTypes.Member
+            }
+
+            for ev in batch.events:
+                if ev.type == EventTypes.Member:
+                    existing_members[ev.state_key] = True
```
[the remark this answers is not in the document]
Comment (ara4n) [reply synapse/handlers/sync.py]: the assumption is that the client will already know about joined members who are now leaving.

File: synapse/handlers/sync.py [hunk 9]
```diff
@@ -494,10 +495,125 @@ def get_state_at(self, room_id, stream_position, types=None, filtered_types=None
...
+            else:
+                summary['m.heroes'] = sorted(
+                    [user_id for user_id in joined_user_ids if user_id != me]
+                )[0:5]
+
+            # ensure we send membership events for heroes if needed
+            cache_key = (sync_config.user.to_string(), sync_config.device_id)
+            cache = self.lazy_loaded_members_cache.get(cache_key)
+
+            existing_members = {
```
[the remark this answers is not in the document]
Comment (ara4n) [reply synapse/handlers/sync.py]: hopefully state won't be that large, as it'll be LL'd

[the remark this answers is not in the document]
Comment (ara4n) [reply synapse/handlers/sync.py]: have added some comments to try to clarify what's going on here.

**Comment by ara4n** [comment]
@richvdh PTAL.

File: synapse/handlers/sync.py [hunk 10]
```diff
@@ -494,10 +495,128 @@ def get_state_at(self, room_id, stream_position, types=None, filtered_types=None
...
+            if ev.content.get("membership") == Membership.JOIN:
+                joined_user_ids.append(ev.state_key)
+            elif ev.content.get("membership") == Membership.INVITE:
+                invited_user_ids.append(ev.state_key)
+
+        # TODO: only send these when they change.
+        summary["m.joined_member_count"] = len(joined_user_ids)
+        summary["m.invited_member_count"] = len(invited_user_ids)
+
+        if not name_id and not canonical_alias_id:
```
Comment (richvdh) [inline synapse/handlers/sync.py]: can we do an early return if `name_id or canonical_alias_id` rather than having this massive `if` block?

File: synapse/handlers/sync.py [hunk 11]
```diff
@@ -494,10 +495,128 @@ def get_state_at(self, room_id, stream_position, types=None, filtered_types=None
...
+                        for user_id in (joined_user_ids + invited_user_ids)
+                        if user_id != me
+                    ]
+                )[0:5]
+            else:
+                summary['m.heroes'] = sorted(
+                    [user_id for user_id in member_ids.keys() if user_id != me]
+                )[0:5]
+
+            if sync_config.filter_collection.lazy_load_members():
```
Comment (richvdh) [inline synapse/handlers/sync.py]: Again, make this into an early return?

**Comment by ara4n** [comment]
I'm very dubious about the early returns; it feels much less error prone to say "only do this thing if this condition is true" rather than "return early if this condition is false, as coincidentally the remainder of the method is handling that condition being true" - as:
 * the latter results in condition-specific stuff being grouped together (indentationwise) with the generic bits of the method
 * it's harder to read, as the reader has to skim for early returns
 * if folks need to add more functionality to the method, they can no longer do it at the end, as the end has ended up on some obscure conditional branch (which may not be obvious unless they skim the whole thing for early returns)
 * it feels like trying to reduce nesting for the sake of reducing nesting.

However, I've gone and done it anyway in a bid to get this merged, but i am still wondering whether we are making reviews too subjective (or making it too hard to determine the 'strength' of review feedback - and whether it's a "if i were doing this, i'd have considered doing it like this, but it's pretty subjective" or a "this is cosmetic nitpick but something that Should Be Fixed.")

**Comment by richvdh** [comment]
I already LGTMed this; it now looks even better. Just waiting for the tests to run.

To address your points, certainly a better alternative might be to split out separate functions. 

> the latter results in condition-specific stuff being grouped together (indentationwise) with the generic bits of the method

yes it does; I don't really see that as a problem.

> it's harder to read, as the reader has to skim for early returns

Personally I find early returns *much* easier to read than having to scroll up and down to figure out how the indentation matches up and which "if" clause we're leaving.

> if folks need to add more functionality to the method, they can no longer do it at the end, as the end has ended up on some obscure conditional branch (which may not be obvious unless they skim the whole thing for early returns)

I think adding stuff to the end of a function without looking at what else is going on would be, er, foolhardy.

> it feels like trying to reduce nesting for the sake of reducing nesting.

sure it is.

>  i am still wondering whether we are making reviews too subjective (or making it too hard to determine the 'strength' of review feedback - and whether it's a "if i were doing this, i'd have considered doing it like this, but it's pretty subjective" or a "this is cosmetic nitpick but something that Should Be Fixed.")

well, we discussed this in the team meeting the other day, and yes, these topics were discussed. I don't think here is the right place to reiterate that conversation, but in this case: I am coming to this as the person who is probably going to have to maintain this code (or at least review future changes to it in the future), and I do not want to have to maintain code that has massive nested if blocks that are hard to follow. Thank you for changing it.
